hivemind/client/averaging/training.py

```diff
@@ -46,27 +48,51 @@ def __init__(self, opt: torch.optim.Optimizer, *, average_parameters: bool, aver
+    def step(self, data_lock: Optional[Lock] = None, wait: bool = True, **kwargs):
+        """ Average optimizer weights and gradients with peers.
+        :param data_lock: averager locks it when model parameters are modified. Otherwise it's assumed that no model
+        modifications occur during averaging step
+        :param wait: if True waits, otherwise returns Future
+        """
         if not wait:
-            return run_in_background(self.step, wait=True, **kwargs)
+            return run_in_background(self.step, data_lock, wait=True, **kwargs)
+
+        # if data_lock is supplied, tensors might change during averaging, so we need to copy them
+        use_old_local_tensors = data_lock is not None
+        if data_lock is None:
+            data_lock = nullcontext()
 
         local_tensors = list(self.local_tensors())
         with self.lock_averager_step:
-            # fill averager's tensors with current local tensors, scaled by peer's weight
-            with self.get_tensors() as averaged_tensors:
-                assert len(local_tensors) == len(averaged_tensors)
+            # fill averager's tensors with current local tensors
+            with data_lock, self.get_tensors() as averaged_tensors:
+                if use_old_local_tensors:
+                    old_local_tensors = tuple(x.cpu().float().clone() for x in local_tensors)
+                assert len(local_tensors) == len(
+                    averaged_tensors), "The number of optimized parameters should not change."
                 for averaged_tensor, local_tensor in zip(averaged_tensors, local_tensors):
-                    averaged_tensor[...] = local_tensor.detach().cpu().float()
+                    averaged_tensor[...] = local_tensor.cpu().float()
 
-            # find a group and hopefully average tensors with peers
+            # find a group and hopefully average tensors with peers, scaled by peer's weight
             gathered = super().step(**kwargs)
-
-            # load averaged tensors back into model
-            with self.get_tensors() as averaged_tensors:
-                assert len(averaged_tensors) == len(local_tensors)
-                for averaged_tensor, local_tensor in zip(averaged_tensors, local_tensors):
-                    local_tensor[...] = averaged_tensor.to(dtype=local_tensor.dtype, device=local_tensor.device)
+            if gathered is not None:
+                # load averaged tensors back into model
+                with data_lock, self.get_tensors() as averaged_tensors:
+                    assert len(averaged_tensors) == len(
+                        local_tensors), "The number of optimized parameters should not change."
+
+                    if use_old_local_tensors:
+                        # since tensors might have changed, we subtract old_local_tensor and add averaged. This prevents
+                        # from loosing local updates that might have occurred during averaging
```

Comment:
```suggestion
                        # since tensors might have changed, we subtract old_local_tensor and add averaged. This prevents
                        # losing local updates that might have occurred during averaging
```